design-documents/0006-metadata-api/text.md

```diff
@@ -0,0 +1,118 @@
+
+[Curators may privately revise a published collection](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/36) with emphasis on [Curators may update published datasets in the private revision of a published collection](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/141) and [Curators may delete published datasets in a private revision for a public collection](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/138)
+
+## Detailed Design | Architecture | Implementation
+
+Summary:
+Create an alias tier between the Explorer and the Data Portal to allow for more flexibility in linkage between a url and a dataset. The explorer will use GET `datasets/meta?url={cellxgene_url}` to retrieve the collection id for the dataset as well as a link to where in S3 the dataset to display is stored. This will require changes in both the portal and the explorer code base.
+
+### Portal Changes
+
+- Add an endpoint to the portal to look up the given url in the explorer_url field for datasets linked to both private and public collections and return
+
+  - collection_id
+  - explorer_s3_uri
+    - Note: If the explorer_url field is not present in the database, return an S3_uri computed based on the explorer_url
+  - Dataset tombstone status
+  - Contact information (for withdrawn by contact message on deleted datasets)
+
+- Add field explorer_s3_uri to the dataset table (string)
+- Modify field deployment_directories (JSON) to be explorer_url (string) and index this field. (This will also require updating code throughout the data portal (front and backend) and the uploader where this field is referenced)
+- Update creation of a dataset revision to create a separate explorer_url
+- Update revision publication to tombstone deleted datasets, update explorer_url for refreshed datasets
+
+For performance the explorer_s3_uri will be cached with a ttl measured in minutes to avoid unnecessary api calls but still allow updates to a dataset to quickly propagate out.
+
+### Explorer Backend Changes
+
+- Add logic to use the portal endpoint to get the cxg url based on the dataset identifier (rather than using the route information and the url to get the dataset directly from S3)
+- Pass the cxg url to the DataAdaptor and follow the current logic patterns.
+
+### Authentication/Authorization
+
+Open question - will this endpoint be available to everyone or only the cellxgene application? If external users have access to the endpoint we will want to consider long term backward compatibility. Additionally we will want to ensure that the no private information is available through the route (or that we require authentication -- creating special permissions for the explorer)
```

Comment:
I think that the endpoint would need to return a subset of the metadata that is currently returned in the _everything but the kitchen sink_ data portal v1 API. For example, this could become the v2 API for public use that only returns metadata useful for the consumer rather than the portal. Currently, the v1 API requires filtering down to a small set of relevant metadata fields.